fix!: check if tx is still valid before including it in a proposal

evan-forbes opened the pull request:
## Overview

fixes #4961 
fixes https://github.com/celestiaorg/celestia-app/issues/4981
maybe also #4824 :crossed_fingers: 

~don't merge until go-square is released and this go.mod is updated~

evan-forbes commented on .goreleaser.yaml:
```diff
@@ -9,10 +9,10 @@ before:
-    - ./scripts/download_v3_binary.sh celestia-app_Linux_arm64.tar.gz celestia-app_linux_v3_arm64.tar.gz v3.10.1-arabica
-    - ./scripts/download_v3_binary.sh celestia-app_Darwin_x86_64.tar.gz celestia-app_darwin_v3_amd64.tar.gz v3.10.1-arabica
-    - ./scripts/download_v3_binary.sh celestia-app_Linux_x86_64.tar.gz celestia-app_linux_v3_amd64.tar.gz v3.10.1-arabica
+    - ./scripts/download_v3_binary.sh celestia-app_Darwin_arm64.tar.gz celestia-app_darwin_v3_arm64.tar.gz v3.10.1-mocha
+    - ./scripts/download_v3_binary.sh celestia-app_Linux_arm64.tar.gz celestia-app_linux_v3_arm64.tar.gz v3.10.1-mocha
+    - ./scripts/download_v3_binary.sh celestia-app_Darwin_x86_64.tar.gz celestia-app_darwin_v3_amd64.tar.gz v3.10.1-mocha
+    - ./scripts/download_v3_binary.sh celestia-app_Linux_x86_64.tar.gz celestia-app_linux_v3_amd64.tar.gz v3.10.1-mocha
```
I originally forked v4.0.2-release, which has this change and main doesn't, however we should bump it on main anyway

evan-forbes commented on app/validate_txs.go:
```diff
@@ -31,46 +63,40 @@ func separateTxs(_ client.TxConfig, rawTxs [][]byte) ([][]byte, []*tx.BlobTx) {
-	blobTxs, _ = filterBlobTxs(logger, txConfig.TxDecoder(), ctx, handler, blobTxs)
-	return append(normalTxs, encodeBlobTxs(blobTxs)...)
-}
+func (fb *FilteredBuilder) Fill(txs [][]byte) [][]byte {
```
the logic here is almost identical to the two functions we had before

however now it includes adding and reverting txs to the square.Builder

evan-forbes commented on internal/embedding/data.go:
```diff
@@ -9,7 +9,7 @@ import (
 // NOTE: This version must be updated at the same time as the version in the
 // Makefile.
-const v3Version = "v3.10.1-arabica"
+const v3Version = "v3.10.1-mocha"
```
#4921

rootulp replied:
Related: https://github.com/celestiaorg/celestia-app/issues/4721

evan-forbes commented on test/docker-e2e/go.sum:
```diff
@@ -350,8 +350,8 @@ github.com/celestiaorg/cosmos-sdk v1.29.2-sdk-v0.50.12 h1:Q5Y9mZymbQP9qUr/l5IPcg
 github.com/celestiaorg/cosmos-sdk/x/upgrade v0.1.0/go.mod h1:T4K9O18zQNKNpt4YvTL3lcUt4aKOEU05ZIFWVdQi3Ak=
-github.com/celestiaorg/go-square/v2 v2.2.0 h1:zJnUxCYc65S8FgUfVpyG/osDcsnjzo/JSXw/Uwn8zp4=
-github.com/celestiaorg/go-square/v2 v2.2.0/go.mod h1:j8kQUqJLYtcvCQMQV6QjEhUdaF7rBTXF74g8LbkR0Co=
+github.com/celestiaorg/go-square/v2 v2.2.1-0.20250612221901-9fc8d4790a54 h1:yyBbRo3LAiu0Pl2ffQmTyXkQLGYvB2UsZwkOHzXDRLA=
+github.com/celestiaorg/go-square/v2 v2.2.1-0.20250612221901-9fc8d4790a54/go.mod h1:6M2txj0j6dkoE+cgwyG0EqrEPhbZpM2R1lsWEopMIBc=
```
note to self to revert before merging

evan-forbes commented on test/util/malicious/out_of_order_prepare.go:
```diff
@@ -46,11 +46,23 @@ func (a *App) OutOfOrderPrepareProposal(req *abci.RequestPrepareProposal) (*abci
 
-	txs := app.FilterTxs(a.Logger(), sdkCtx, handler, a.GetTxConfig(), req.Txs)
+	fb, err := app.NewFilteredBuilder(
+		a.Logger(),
+		sdkCtx,
+		handler,
+		a.GetEncodingConfig().TxConfig,
+		a.MaxEffectiveSquareSize(sdkCtx),
+		appconsts.SubtreeRootThreshold,
+	)
+	if err != nil {
+		panic(err)
+	}
+
+	txs := fb.Fill(req.Txs)
 
 	// build the square from the set of valid and prioritised transactions.
 	// The txs returned are the ones used in the square and block
-	dataSquare, txs, err := Build(txs, appconsts.LatestVersion, a.MaxEffectiveSquareSize(sdkCtx), OutOfOrderExport)
+	dataSquare, err := fb.Build()
```
really hope we can just delete this package as maintaining two copies of things is cursed #4990

damiannolan commented on app/validate_txs.go:
```diff
@@ -8,11 +8,44 @@ import (
 
+// FilteredBuilder filters txs and blobs using a copy of the state and tx validity
+// rules before adding it the square.
+type FilteredBuilder struct {
```
maybe we could call this `FilteredSquareBuilder` to be more specific about what the type is used for, I think it would help the reader

I would also suggest moving the code in this file (`validate_txs.go` -> `square.go` or something similar). I think `square_size.go` exists in this pkg with a single small func which could maybe be moved and live next to all things square 🟦 (although, I think its a method on the app)

damiannolan commented on app/validate_txs.go:
```diff
@@ -8,11 +8,44 @@ import (
+// FilteredBuilder filters txs and blobs using a copy of the state and tx validity
+// rules before adding it the square.
+type FilteredBuilder struct {
+	logger   log.Logger
+	ctx      sdk.Context
```
I think we could remove these as fields on the struct and pass `ctx` to `Fill(ctx, req.Txs)`, and any other funcs as necessary

Also could derive a logger from ctx on the fly using something like:
```
logger := ctx.Logger().With("app/filtered-square-builder")
```

damiannolan commented on app/test/prepare_proposal_test.go:
```diff
@@ -30,6 +33,74 @@ import (
+		require.NoError(t, err)
+		require.Equal(t, 7, len(prepareResponse.Txs))
+
+		fmt.Println("using height", height)
```
using `fmt` might flood test output unncessarily

```suggestion
		t.Logf("using height %d\n", height)
```

rootulp replied:
Also this log doesn't seem helpful after the test was written

```suggestion
```

damiannolan reviewed (commented):
lgtm! I just left some nitty comments about code org and structuring which I think might help non-familiar devs

rootulp commented on app/test/prepare_proposal_test.go:
```diff
@@ -30,6 +33,74 @@ import (
 
+func TestPrepareProposalValidConstruction(t *testing.T) {
+	// Reproduces https://github.com/celestiaorg/celestia-app/issues/4961
+	t.Run("prepare proposal creates a proposal that process proposal throws an account sequence mismatch", func(t *testing.T) {
```
[nit] now that the test passes 🎊 

```suggestion
	t.Run("prepare proposal creates a proposal that process proposal accepts", func(t *testing.T) {
```

rootulp commented on app/test/prepare_proposal_test.go:
```diff
@@ -30,6 +33,74 @@ import (
+			Height: height,
+			Time:   time.Now(),
+		})
+		require.NoError(t, err)
+		require.Equal(t, 7, len(prepareResponse.Txs))
```
[optional]

```suggestion
		require.NoError(t, err)
		// The filtered builder in prepare proposal should have dropped the last two txs.
		// The filtered builder should have dropped the second to last tx because it was too large to fit in the square.
		// The filtered builder should have dropped the last tx because the nonce for it was invalidated by dropping the second to last tx.
		require.Equal(t, 7, len(prepareResponse.Txs))
```

rootulp commented on app/validate_txs.go:
```diff
@@ -8,11 +8,44 @@ import (
+	maxSquareSize,
+	subtreeRootThreshold int,
+) (*FilteredBuilder, error) {
+	builder, err := square.NewBuilder(maxSquareSize, subtreeRootThreshold)
```
[nit] it can be tricky to debug functions that return a non nil value AND an error. IMO for new code, we should return value or error but never both.

```suggestion
	builder, err := square.NewBuilder(maxSquareSize, subtreeRootThreshold)
	if err != nil {
		return nil, err
	}
```

rootulp commented on app/validate_txs.go:
```diff
@@ -31,101 +64,106 @@ func separateTxs(_ client.TxConfig, rawTxs [][]byte) ([][]byte, []*tx.BlobTx) {
 
 		msgTypes := msgTypes(sdkTx)
 		if nonPFBMessageCount+len(sdkTx.GetMsgs()) > appconsts.MaxNonPFBMessages {
-			logger.Debug("skipping tx because the max non PFB message count was reached", "tx", tmbytes.HexBytes(coretypes.Tx(tx).Hash()))
+			fb.logger.Debug("skipping tx because the max non PFB message count was reached", "tx", tmbytes.HexBytes(coretypes.Tx(tx).Hash()))
 			continue
 		}
```
[blocking][question] the builder already appended the tx on line 85 but here we're skipping running the handler on it and we haven't reverted it from the builder. I would expect this to call

```go
		if nonPFBMessageCount+len(sdkTx.GetMsgs()) > appconsts.MaxNonPFBMessages {
			fb.logger.Debug("skipping tx because the max non PFB message count was reached", "tx", tmbytes.HexBytes(coretypes.Tx(tx).Hash()))
			err = fb.builder.RevertLastTx()
			if err != nil {
				fb.logger.Error("reverting last transaction", "error", err)
			}
			continue
		}
```

rootulp replied:
Alternatively this bug isn't present in the blob txs section below because appending happened after this message count check so we could move it down like so:

```go
		msgTypes := msgTypes(sdkTx)
		if nonPFBMessageCount+len(sdkTx.GetMsgs()) > appconsts.MaxNonPFBMessages {
			logger.Debug("skipping tx because the max non PFB message count was reached", "tx", tmbytes.HexBytes(coretypes.Tx(tx).Hash()))
			continue
		}


		if !fb.builder.AppendTx(tx) {
			continue
		}
```

rootulp commented on test/util/malicious/out_of_order_prepare.go:
```diff
@@ -46,11 +46,21 @@ func (a *App) OutOfOrderPrepareProposal(req *abci.RequestPrepareProposal) (*abci
 	// build the square from the set of valid and prioritised transactions.
 	// The txs returned are the ones used in the square and block
-	dataSquare, txs, err := Build(txs, appconsts.LatestVersion, a.MaxEffectiveSquareSize(sdkCtx), OutOfOrderExport)
+	dataSquare, err := fb.Build()
```
Disregard. `fb.Build()` doesn't prune any txs so the txs from above are still valid.

rootulp commented:
nit: I think we can drop the `!` from the PR title b/c doesn't seem consensus breaking.

evan-forbes commented:
okay, just fixed the malicious test in https://github.com/celestiaorg/celestia-app/pull/4985/commits/d8723ad771cf70919316611a44234a06402ff57f :facepalm: so we should be good to merge!

> I think we can drop the ! from the PR title b/c doesn't seem consensus breaking.

per sync, I think this is technically (but not meaningfully) api breaking since we deleted `FilterTxs`. imo, it doesn't make sense to go through the normal deprecation process since it is only used by this repo

rootulp commented:
> However, I am missing where the transactions get rejected if they're invalid? I'm referring to the case where two transactions are from the same account, if the former gets rejected, and the latter has a higher nonce, the latter should also be rejected. Where is that happening?

It is happening inside the ante handler (`handler`)
